File: T9/Trie.m
//
//  Trie.m
//  T9
//
//  Created by Bernhard Bauer on 19.05.06.
//  Copyright 2006 Black Sheep Software. All rights reserved.
//

#import "Trie.h"

#include <ctype.h>

static inline int indexForChar(char c) {
	return tolower(c)-'a';
}

static inline int charForIndex(int i) {
	return 'a'+i;
}

@implementation Trie

- (id)init {
	if (self = [super init]) {
		int i;
		for (i=0; i<NUM_CHARS; i++) {
			children[i] = nil;
		}
		freq = freqSum = 0;
	}
	return self;
}

- (id)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];
	int i;
	freqSum = freq = [coder decodeIntForKey:@"T9Freq"];
	for (i=0; i<NUM_CHARS; i++) {
		Trie* child = [coder decodeObjectForKey:[NSString stringWithFormat:@"T9Child%c", charForIndex(i)]];
		children[i] = [child retain];
		freqSum += [child frequencySum];
	}
	return self;
}

- (void)dealloc {
	int i;
	for (i=0; i<NUM_CHARS; i++) {
		[children[i] release];
		children[i] = nil;
	}
	[super dealloc];
}
	

- (void)encodeWithCoder:(NSCoder*)coder {
	[super encodeWithCoder: coder];
	[coder encodeInt:freq forKey:@"T9Freq"];
	int i;
	for (i=0; i<NUM_CHARS; i++) {
		[coder encodeObject:children[i] forKey:[NSString stringWithFormat:@"T9Child%c", charForIndex(i)]];
	}
}

- (unsigned int)frequency {
	return freq;
}

- (unsigned int)frequencySum {
	return freqSum;
}

- (Trie*)childForChar:(char)c {
	return children[indexForChar(c)];
}

- (void)learnNgramFrom:(char*)start to:(char*)end {
	freqSum++;
	if (start > end) {
		freq++;
		return;
	}
	int index = indexForChar(*end);
	Trie* child = children[index];
	if (child == nil) {
		children[index] = [[Trie alloc] init];
	}
	[child learnNgramFrom:start to:(end-1)];
}

- (void)learnString:(NSString*)s {
	const char* buffer = [s cString];
	int i;
	int length = [s cStringLength];
	for (i=-1; i<length; i++) {
		[self learnNgramFrom:buffer to:(buffer+i)];
	}
}

@end

File: src/main/java/de/blacksheepsoftware/t9/Trie.java
package de.blacksheepsoftware.t9;

import java.io.*;

public class Trie implements Serializable {
	
	private static final long serialVersionUID = -5865474431091083665L;

	public static final int NUM_CHARS = 26;
	
	protected long freq = 0;
	volatile protected long freqSum = 0;
	
	protected Trie[] children = new Trie[NUM_CHARS];
	

	public long frequency() {
		return freq;
	}
	
	public long frequencySum() {
		if (freqSum == 0) {
			freqSum = freq;
			for (int i=0; i<NUM_CHARS; i++) {
				Trie child = children[i];
				if (child != null) freqSum += child.frequencySum();
			}
		}
		return freqSum;
	}
	
	public Trie child(char c) {
		return children[Character.getNumericValue(c)-10];
	}
	
	public Trie child(Character c) {
		return children[Character.getNumericValue(c.charValue())-10];
	}
	
	public void learn(String s, int i) {
		freqSum++;
		if (i < 0) {
			freq++;
			return;
		}
		int childIndex = Character.getNumericValue(s.charAt(i))-10;
		if (children[childIndex] == null) {
			children[childIndex] = new Trie();
		}
		children[childIndex].learn(s, i-1);
	}
	
	public void learn(String s) {
		for (int i=-1; i<s.length(); i++) {
			learn(s, i);
		}
	}
}
